central/image/service/service_impl.go

```diff
@@ -476,9 +480,17 @@ func (s *serviceImpl) GetImageVulnerabilitiesInternal(ctx context.Context, reque
 	return internalScanRespFromImage(img), nil
 }
 
+func (s *serviceImpl) releaseFromScanSemaphore() {
+	s.internalScanSemaphore.Release(1)
+	images.ScanSemaphoreHoldingSize.With(imageScanMetricsLabel).Dec()
+}
+
 func (s *serviceImpl) acquireScanSemaphore(ctx context.Context) error {
 	semaphoreCtx, cancel := context.WithTimeout(ctx, maxSemaphoreWaitTime)
 	defer cancel()
+	images.ScanSemaphoreQueueSize.With(imageScanMetricsLabel).Inc()
+	defer images.ScanSemaphoreQueueSize.With(imageScanMetricsLabel).Dec()
+	defer images.ScanSemaphoreHoldingSize.With(imageScanMetricsLabel).Inc()
```

Comment:
I am afraid, I do not fully understand the case here. No matter what happens later (due to `defer`), we are already now decreasing the queue size and increase the holding size?
And what happens if `s.internalScanSemaphore.Acquire(semaphoreCtx, 1)` returns an error and the function exits? The number would be incorrect, right?

Reply:
I changed the logic in [4c91c8c](https://github.com/stackrox/stackrox/pull/14695/commits/4c91c8c3a88bc31d779ed09141fb194f05cba360) - resolving.